components/places/src/storage.rs

```diff
@@ -263,21 +270,311 @@ fn add_visit(db: &impl ConnExt,
+        // Option<bool> - None vs Some(false) is confusing. We should use an enum.
+        update_frecency(&db, page_info.row_id, None)?;
+
+        // and the place itself if necessary.
+        let new_title = title.as_ref().unwrap_or(&page_info.title);
+        // We set the Status to Normal, otherwise we will re-upload it as
+        // outgoing even if nothing has changed.
+        db.execute_named_cached(
+            "UPDATE moz_places
+             SET title = :title,
+                 sync_status = :status
+             WHERE id == :row_id",
+            &[(":title", new_title),
+              (":row_id", &page_info.row_id),
+              (":status", &SyncStatus::Normal)]
+        )?;
+
+        // and finally update the guid if necessary.
+        // XXX - we should consider doing this as part of the update above, but
+        // it complicates the code and should be rare.
+        if let Some(ref existing_guid) = existing_guid {
+            if existing_guid != new_guid {
+                db.execute_named_cached(
+                    "UPDATE moz_places SET guid = :new_guid WHERE guid == :old_guid",
+                    &[(":new_guid", new_guid), (":old_guid", existing_guid)]
+                )?;
+            }
+        }
+        Ok(())
+    }
+
+    pub fn apply_synced_reconcilliation(db: &Connection, guid: &SyncGuid) -> Result<()> {
+        db.execute_named_cached(
+            "UPDATE moz_places
+             SET sync_status = :status,
+                 sync_change_counter = 0
+             WHERE guid == :guid",
+            &[(":guid", guid),
+              (":status", &SyncStatus::Normal)]
+        )?;
+        Ok(())
+    }
+
+    pub fn apply_synced_deletion(db: &Connection, guid: &SyncGuid) -> Result<()> {
+        db.execute_named_cached(
+            "DELETE FROM moz_places WHERE guid = :guid",
+            &[(":guid", guid)])?;
+        // This might have created a local tombstone for the item, which we
+        // don't want.
+        db.execute_named_cached(
+            "DELETE FROM moz_places_tombstones WHERE guid = :guid",
+            &[(":guid", guid)])?;
+        Ok(())
+    }
+
+    #[derive(Debug)]
+    pub enum OutgoingInfo {
+        Record(HistoryRecord),
+        Tombstone,
+    }
+
+    pub fn fetch_outgoing(db: &Connection,
+                          max_places: usize,
+                          max_visits: usize) -> Result<HashMap<SyncGuid, OutgoingInfo>> {
+        // Note that we want *all* "new" regardless of change counter,
+        // so that we do the right thing after a "reset".
+        let mut stmt = db.conn().prepare(&format!("
+            SELECT guid, url, id, title, hidden, typed, frecency,
+                visit_count_local, visit_count_remote,
+                last_visit_date_local, last_visit_date_remote,
+                sync_status, sync_change_counter
+            FROM moz_places
+            WHERE (sync_change_counter > 0 OR sync_status != {})
```

Comment:
Ah, this answers my question from earlier. Why the disjunction, though? Adding visits bumps the change counter, so I think just `sync_change_counter > 0` is sufficient?